consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ProposalValidator.java

```diff
@@ -0,0 +1,288 @@
+      final Optional<SignedData<RoundChangePayload>> roundChangeWithLatestPreparedRound =
+          getRoundChangeWithLatestPreparedRound(proposal.getRoundChanges());
+
+      if (roundChangeWithLatestPreparedRound.isPresent()) {
+        final PreparedRoundMetadata metadata =
+            roundChangeWithLatestPreparedRound.get().getPayload().getPreparedRoundMetadata().get();
+
+        // The Hash in the roundchange/proposals is NOT the same as the value in the
+        // prepares/roundchanges
+        // as said payloads reference the block with an OLD round number in it - therefore, need
+        // to create a block with the old round in it, then re-calc expected hash
+        // Need to check that if we substitute the LatestPrepareCert round number into the supplied
+        // block that we get the SAME hash as PreparedCert.
+        final Block currentBlockWithOldRound =
+            BftBlockInterface.replaceRoundInBlock(
+                proposal.getBlock(),
+                metadata.getPreparedRound(),
+                BftBlockHeaderFunctions.forCommittedSeal());
+
+        final Hash expectedPriorBlockHash = currentBlockWithOldRound.getHash();
+
+        if (!metadata.getPreparedBlockHash().equals(expectedPriorBlockHash)) {
+          LOG.info(
+              "{}: Latest Prepared Metadata blockhash does not align with proposed block",
+              ERROR_PREFIX);
+          return false;
+        }
+
+        // validate the prepares
+        if (!validatePrepares(
+            metadata, proposal.getRoundIdentifier().getSequenceNumber(), proposal.getPrepares())) {
+          LOG.info("{}: Piggy-backed prepares failed validation", ERROR_PREFIX);
+          return false;
+        }
+      } else {
+        // no one prepared, so prepares should be empty
+        if (!proposal.getPrepares().isEmpty()) {
+          LOG.info("{}: No PreparedMetadata exists, so prepare list must be empty", ERROR_PREFIX);
+          return false;
+        }
+
+        return validateBlockCoinbaseMatchesMsgAuthor(proposal);
+      }
+
+      return true;
+    }
+  }
+
+  private boolean validateRoundZeroProposalHasNoRoundChangesOrPrepares(final Proposal proposal) {
+    if ((proposal.getRoundChanges().size() != 0) || proposal.getPrepares().size() != 0) {
+      LOG.info("{}: round-0 proposal must not contain any prepares or roundchanges", ERROR_PREFIX);
+      return false;
+    }
+
+    return true;
+  }
+
+  private boolean validateRoundChanges(
+      final Proposal proposal, final List<SignedData<RoundChangePayload>> roundChanges) {
+
+    if (hasDuplicateAuthors(roundChanges)) {
+      LOG.info("{}: multiple round changes from the same author.", ERROR_PREFIX);
+      return false;
+    }
+
+    if (!hasSufficientEntries(roundChanges, quorumMessageCount)) {
+      LOG.info("{}: Insufficient round changes for proposal", ERROR_PREFIX);
+      return false;
+    }
+
+    if (!metadataIsConsistentAcrossRoundChanges(roundChanges)) {
```

Comment:
Does `metadataIsConsistentAcrossRoundChanges` verify that all RoundChanges that have the same preparedRound also have the same value of preparedBlockHash?

If so, then this is not correct.

Reply:
What I mean is that, if so, then the current validation fails valid Proposal messages.

Reply:
Yes, this is what this validation step. It ensures that all RoundChanges which have prepared-Round have the same blockhash - other than Byzantine behaviour, is there a way this could be possible?
And, if this detects byzatine behaviour, then I had assumed the packet should be deemed invalid.

In general, we declare a packet invalid very quickly - and this logic COULD be changed, i.e. rather than throwing the whole "parent" packet away due to a poor piggy-back, we COULD simply filter out the bad ones, then work with the remainder. However, that becomes harder across the full gamut of the implementation.